### DISCUSSION TITLE
[WIP: 2024 Update] C# wrappers

### DISCUSSION THREAD

**Discussion by GeorgeS2019** [rdkit/rdkit#7460] [Development]
@jones-gareth
I understand you provide the latest c# wrapper. Is there other way to access the artifact except the Nuget?

> This discussion aims eventually to have a **CI/CD** artifact for c# wrapper as Nuget
### Release_2024.03.3
Handle query atoms and bonds in SWIG wrappers
https://github.com/rdkit/rdkit/pull/7431

#### How to add MolStandardize to C# wrappers
https://github.com/rdkit/rdkit/pull/7351/

#### C# Build Net6 library and tests using cmake
https://github.com/rdkit/rdkit/pull/7326

#### Add MolEnumerator to C#
https://github.com/rdkit/rdkit/pull/7351

#### Fix SWIG wrappers for C#
https://github.com/rdkit/rdkit/pull/4935

#### Fix SWIG wrappers for C#
https://github.com/rdkit/rdkit/pull/4935

#### Fix SWIG wrappers for C#
https://github.com/rdkit/rdkit/pull/4935

**Comment by GeorgeS2019** [comment]
kind of new here, all python codes here are not exposed for c# wrapper
e.g. https://github.com/rdkit/rdkit/tree/master/rdkit/Chem/Draw

**Reply by jones-gareth** [reply]
Both the `Java` and `C#` wrappers wrap `C++` code using [SWIG](https://swig.org/).  SWIG is unable to wrap Python.

So the portions of the RDKit written in Python, such as the IPython drawing extension you found, cannot be wrapped. There is other stuff written in Python such as the molecular registration hashes and some ML code.

You are able to write PNG and SVG images of molecules using the `C#` wrapper.

**Comment by jones-gareth** [comment]
> @jones-gareth I understand you provide the latest c# wrapper. Is there other way to access the artifact except the Nuget?
> 
> > This discussion aims eventually to have a **CI/CD** artifact for c# wrapper as Nuget
> 

While I may have been the most active recently in the C# wrappers, most of the work has been done by others previously.  

Outside of Nuget you could build the wrappers yourself from source.  There is an RDKit test that uses the Azure pipeline at [vs_build_swig.yaml](https://github.com/rdkit/rdkit/blob/master/.azure-pipelines/vs_build_swig.yml).  If you execute the commands in the pipeline you can build the C# wrapper locally.

Not sure what you mean by "have a **CI/CD** artifact for c# wrapper as Nuget". The [vs_build_swig.yaml](https://github.com/rdkit/rdkit/blob/master/.azure-pipelines/vs_build_swig.yml) check builds the C# wrapper and runs the (admittedly small) number of C# tests. I imagine it could be extended to create a `nupkg` file, but I don't think you would want to publish from within the test.
